Add more metrics to categorical.py, modified the wrappers.py the support Weibull’s plotting position and add new file implement_categorical_metrics.ipynb to provide tutorial

LQscience opened the pull request:
# Modified:   weatherbenchX/metrics/categorical.py
The following metrics are added: 
- Symmetric extreme dependence score (SEDS)
- Symmetric extremal dependence index (SEDI)
- False alarm rate (F), probability of false detection
- Heidke skill score (HSS), Cohen’s Kappa 
- Odds ratio skill score (ORSS), Yule’s Q
- Extreme dependence score (EDS)
- Extremal dependence index (EDI)
- Gilbert skill score (GSS), equitable threat score
- Peirce’s skill score (PSS), Hanssen and Kuipers discriminant
- Specificity, true negative rate (TNR)
- Area under receiver operating characteristic (ROC) curve (AUC)
- ROC skill score (ROCSS)
- Relative economic value (REV)

# Modified:   weatherbenchX/metrics/wrappers.py
- modified the function "binarize_thresholds" to support threshold of Xarray.DataArray
- Add new class "EnsembleToProbabilistic" to convert ensemble forecasts into probabilitic forecast using the Weibull’s plotting position (Makkonen, 2006).
Makkonen, L.: Plotting Positions in Extreme Value Analysis, Journal of Applied Meteorology and Climatology, 
      45, 334–340, https://doi.org/10.1175/JAM2349.1, 2006.

# Add new file: docs/source/how_to/implement_categorical_metrics.ipynb
This notebook provides tutorial to use these categorical metrics.
Also, it is a part of code to reproduce the analysis of the manuscript "The extension of the WeatherBench 2 to binary hydroclimatic forecasts" (https://egusphere.copernicus.org/preprints/2025/egusphere-2025-3/) that has been submitted to the Geoscientific Model Development.

raspstephan commented:
Hi,
thank you so much for this PR. I will take a close look tomorrow.

raspstephan commented:
This is really great work and will be a valuable addition to WeatherBench. Generally, the added code looks great. I have a few suggestions though:

1. Since the notebook is relatively specific for your paper, I would suggest adding the notebook as a supplement to the paper instead of adding it to the WB-X documentation. Instead, I can add a section to the documentation called something like "Used by" which lists all example of people using WB-X and add a reference to your paper and notebook there.

2. Adding an option of using a DataArray as a threshold is great and is actually something we were planning to do as well. However, currently this only allows using a DataArray which means that the same thresholds apply to all variables. Some users might want to use different thresholds for different variables though (I think you do as well judging from your notebook.) One way to do this would be to also allow using an xr.Dataset and adding a logic something like (I did not actually test this, just pseudo-code):
```
if isinstance(thresholds, xr.Dataset):
  threshold_for_variable = thresholds[x.name]
```

3. Since this is quite a lot of new functionality and metrics, I would love it if you could add some tests. Specifically for the new xr.DataArray/Dataset options in the wrapper. You can decide for yourself whether you'd like to add tests for the metrics.

raspstephan commented on weatherbenchX/metrics/categorical.py:
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
+    )
+
+
+class F(base.PerVariableMetric):
```
I think it would be better if this was named FalseAlarmRate since F can be a little confusing.

mjwillson replied:
+1. We have a lot of dense acronyms in class names here. I think expanding the acronyms would generally be good for readability, except  in cases where this would lead to a very long name, or where the acronym is very well-known and unambiguous.

So e.g. SEEPS -> StableEquitableErrorInProbabilitySpaceScore would probably be overkill, but things like F -> FalsePositiveRate or PSS -> PierceSkillScore seem like an improvement to me.

LQscience replied:
Sorry for the confusion. I have checked all metrics and modified their names to their full names.

raspstephan commented on weatherbenchX/metrics/categorical.py:
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
+    return ((predictions - targets) ** 2).astype(np.float32)
+
+
+class BrierScore(base.PerVariableMetric):
```
So Brier Score actually doesn't require any additional code since mathematically it's the same as deterministic.MSE(). However, I would be happy to still implement it (but then using the squared error definition from deterministic.py).

But could you explain why you need to add a reduce_dims argument? This should be handled by the aggregator, no?

mjwillson replied:
+1 perhaps just aliasing deterministic.MSE here as BrierScore would be sufficient?

You add a float32 cast here vs in deterministic.SquaredError, but I'm not sure this is necessary if this is wrapped with wrappers.ContinuousToBinary as binarize_thresholds already returns floats, due to the .where: https://github.com/google-research/weatherbenchX/blob/main/weatherbenchX/metrics/wrappers.py#L69. (@raspstephan I think the docs on this are incorrect, AFAICT it always returns floats regardless whether NaNs are present or not?).

Also note that treating Brier score as a special case of MSE allows you to reuse probabilistic.UnbiasedEnsembleMeanSquaredError to get a fair / unbiased estimate of Brier score from an ensemble. (In the special case of Brier score this is equivalent to the unbiased estimator from Ferro, C.A., 2007: Comparing Probabilistic Forecasting Systems with the Brier Score, Eqn 8 with M->inf)

LQscience replied:
The ".astype(np.float32)" is added here because I found that the calculation for global grid cells ususlly be out of memory for my PC. Of course,  it can be deleted.

LQscience replied:
Thank you for your suggestion. I decide to remove this function in this PR and add it in the next PR when I finish to prapare the "metrics_test.py".

raspstephan commented on weatherbenchX/metrics/categorical.py:
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
+    return ((e_clim - e_fcst) / (e_clim - e_perfect)).max(dim=self._threshold_probability_dim)
+
+
+class SquaredError(base.PerVariableStatistic):
```
Same as deterministic.SquaredError, see comment below.

LQscience replied:
Thank you for your suggestion. I decide to remove this function in this PR and add it in the next PR when I complete to prapare the "metrics_test.py".

raspstephan commented on weatherbenchX/metrics/categorical.py:
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
+    )
+
+
+class AUC(base.PerVariableMetric):
```
I think here a test would be great to see if the function returns the expected results in some edge cases.

LQscience replied:
Thank you for your suggestion. I decide to remove this function in this PR and add it in the next PR when I complete to prapare the "metrics_test.py".

mjwillson commented on weatherbenchX/metrics/categorical.py:
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
+    """Computes metrics from aggregated statistics."""
+
+    event_count = statistic_values['TruePositives'] + statistic_values['FalseNegatives']
+    event_nonevent_count = event_count + statistic_values['FalsePositives'] + statistic_values['TrueNegatives']
```
FYI this should always be 1 (up to numerical error) because weatherbench gives you *means* of the statistics here. The entries of the confusion matrix always sum to 1 on a per-example basis, and so their means should sum to 1 as well.

In the case of numerical error I'm actually not 100% sure what's best to do here, but I don't think it would generally make a big difference so perhaps better to go with the simpler option and avoid the extra division?

mjwillson commented on weatherbenchX/metrics/categorical.py:
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
+      ) / event_nonevent_count * self._cost_loss_ratios + statistic_values['FalseNegatives'] / event_nonevent_count
+    e_perfect = event_ratio * self._cost_loss_ratios
+    
+    return ((e_clim - e_fcst) / (e_clim - e_perfect)).max(dim=self._threshold_probability_dim)
```
Most of this duplicates code in REV._values_from_mean_statistics_per_variable, would be nice to factor out into a shared helper.

LQscience replied:
As the PREV can be easily calculated using the max function, i decided to delete the PREV class.

mjwillson commented on weatherbenchX/metrics/categorical.py:
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
+  def __init__(
+      self,
+      cost_loss_ratios: Union[float, Sequence[float]],
+      threshold_probability_dim: str,
```
Could you document a bit on the requirements for this threshold_probability_dim and how it would typically be supplied? I think the idea is that you wrap this with ContinuousToBinary with multiple thresholds set for the probability, right?

LQscience replied:
I have to clarify that the "threshold_probability_dim" here is different with the "threshold_dim" of the "wrappers.ContinuousToBinary". The former is used to convert the probabilistic forecasts into binary forecasts, while the latter is used to convert the continuous forecasts into binary forecasts. For calculating the PREV, we need to predefined a set of probability thresholds over which the maximum REV is calculated.

mjwillson commented on weatherbenchX/metrics/categorical.py:
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
+    hit_rate = statistic_values['TruePositives'] / (
+        statistic_values['TruePositives'] + statistic_values['FalseNegatives']
+    )
+    # print(hit_rate)
```
Please remove, here and below

LQscience replied:
Sorry for the lines not deleted after testing. I have removed them.

mjwillson commented on weatherbenchX/metrics/categorical.py:
```diff
@@ -487,3 +487,717 @@ def _values_from_mean_statistics_with_internal_names(
+        statistic_values['FalsePositives'] + statistic_values['TrueNegatives']
+    )
+    # print(false_alarm_rate)
+    auc_dims = tuple(set(hit_rate.dims) - {self._threshold_probability_dim})
```
This shares most of its code with AUC, could we please factor out the shared code into a helper?

mjwillson replied:
Alternatively I think it would be fair to decide to implement just one of AUC and ROCSS = 2 * AUC - 1, since they can trivially be transformed into eachother by the end-user of the metrics. Maybe just AUC as I think that's better known?

LQscience replied:
Thanks for your suggestion. I have deleted the ROCSS.

mjwillson reviewed (commented):
Thanks, this is a really great contribution! Made a few comments, mostly minor.

mjwillson commented on weatherbenchX/metrics/wrappers.py:
```diff
@@ -395,3 +401,31 @@ def _values_from_mean_statistics_with_internal_names(
+
+class EnsembleToProbabilistic(InputTransform):
+  """
+  Convert ensemble forecasts into probabilitic forecast using the Weibull’s plotting position (Makkonen, 2006).
```
Also good to clarify that you need to binarize forecasts before this (e.g. by first wrapping with ContinuousToBinary)

LQscience replied:
Thank you for the constructive comments. The class name has been modified to "WeibullEnsembleToProbabilistic". In the meantime, the need to binarize the forecasts before this wrapper has been clrified in the introduction.

LQscience commented:
> Thanks, this is a really great contribution! Made a few comments, mostly minor.

Thanks you very much  for the positive feedback and detailed suggestions! I'll check and address all the issues this weekend.

LQscience commented on weatherbenchX/metrics/wrappers.py:
```diff
@@ -395,3 +401,31 @@ def _values_from_mean_statistics_with_internal_names(
+
+  def tranform_fn(self, da: xr.DataArray) -> xr.DataArray:
+    ensemble_members = da[self._ensemble_dim].size
+    return da.sum(self._ensemble_dim, skipna=self._skipna)/(ensemble_members+1)
```
I have realized it is true that the ensemble_members would generally need to be adjusted based on the number of non-NaN ensemble members. So the skipna is expected to be False here and I have modified that.

langmore commented on weatherbenchX/metrics/wrappers.py:
```diff
@@ -63,15 +63,15 @@ def binarize_thresholds(
     binary_x: Binarized DataArray.
   """
-  if not isinstance(thresholds, xr.DataArray):
+  if not isinstance(thresholds, xr.Dataset):
```
In `binarize_thresholds` def, you can update the typing of thresholds to `Iterable[float] | xr.Dataset`

raspstephan commented on ._.DS_Store:
Please remove :)

LQscience replied:
OK, I have already removed it.

raspstephan commented:
Thanks a lot for addressing all the comments. I think the only thing that is remaining is to add tests. I think we definitely want to test the behavior or the wrapper to make sure using an xr.Dataset as a threshold works as expected. It would be great to also have tests for the new metrics but we don't have full test coverage of the existing metrics, so it's not a strict requirement.

LQscience commented:
> Thanks a lot for addressing all the comments. I think the only thing that is remaining is to add tests. I think we definitely want to test the behavior or the wrapper to make sure using an xr.Dataset as a threshold works as expected. It would be great to also have tests for the new metrics but we don't have full test coverage of the existing metrics, so it's not a strict requirement.

Thank you very much for your positive feedback. I have already test the added metrics and will test several other complex metrics in the following few days. After testing, I will submit these complex metrics as well as the test of all the metrics.

langmore commented:
I have a pending (internal) change that adds a file for tests of `InputTransform`. This may be a good home for some of your tests. In particular, I'm encouraging users to test the `InputTransform` *without* any metric.  So, @LQscience if you're writing tests of `InputTransform`'s now, just add them to `metrics_test.py` (we can move later), and test the transform without a metric.

LQscience commented:
> I have a pending (internal) change that adds a file for tests of `InputTransform`. This may be a good home for some of your tests. In particular, I'm encouraging users to test the `InputTransform` _without_ any metric. So, @LQscience if you're writing tests of `InputTransform`'s now, just add them to `metrics_test.py` (we can move later), and test the transform without a metric.

Thanks for the kind reminder! I’ll add the tests to "metrics_test.py" as soon as possible.

raspstephan commented:
Hey @LQscience, just checking in whether you are still working on this? No rush but some of us are quite excited for the features you implemented, so if I can help with anything let me know.

LQscience commented:
> Hey @LQscience, just checking in whether you are still working on this? No rush but some of us are quite excited for the features you implemented, so if I can help with anything let me know.

Dear Rasp,

I sincerely apologize for the delay in my response. I have been preoccupied with other tasks recently, which prevented me from addressing your feedback in a timely manner. I truly appreciate your patience and understanding.

I have now made the following updates based on your feedback:
1. Improve the `binarize_thresholds` and `ContinuousToBinary` functions in the `wrappers.py`to support threshold in the forms of  xarray.DataArray and xarray.Dataset. 
2. Add a new class `WeibullEnsembleToProbabilistic` into the `wrappers.py` for converting ensemble forecasts into probabilistic forecasts using Weibull’s plotting position. 
3. Add tests in the `wrappers_test.py` to validate these changes above.

For more concise and efficient modification and review processes, I will close this lengthy pull request. Please refer to the new [pull request](https://github.com/google-research/weatherbenchX/pull/52) for the changes above. I am currently continuing to improve and test the remaining issues and will make new pull request.

Please let me know if there are any further modifications needed. Thank you for your guidance and support.

Best regards,
Qiang Li

raspstephan commented:
Just out of curiosity, are you planning to send another PR with REV?
